src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
+  deriving (Eq)
+
+instance (Show i, Show e) => Show (ParseErrorType i e) where
+  show :: ParseErrorType i e -> String
+  show EndOfInput = "End of Input"
+  show (Unexpected a) = "Unexpected " ++ show a
+  show (Expected e a) = "Expected " ++ show e ++ ", got " ++ show a
+  show (ExpectedEndOfFile a) = "Expected EndOfFile, got " ++ show a
+  show (CustomError e) = "Custom Error: " ++ show e
+  show Empty = "No error"
+
+--------------------
+---- ParseError ----
+--------------------
+
+type Offset = Integer
+
+data ParseError i e where
+  ParseError
+    :: { errOffset :: Offset
```

Comment:
Unnecessary indentation caused by writing the `::` on a new line.

Reply:
This should improve when I meddle with the `fourmolu` config a bit, stay tuned